### ISSUE TITLE
Discuss about fundamental solution of detecting symlink file based bypass method

### ISSUE THREAD

**Issue by hi120ki** [falcosecurity/falco#2203]
As I mentioned in PR #2202, there is bypass method to `fd.name` match rules by symlink files, and it can bypass many rules.

In latest alpine image, there are these symlink files that can be used for bypass rules.

```
/proc/self/root -> /
/proc/self/cwd -> (proccess dir)
/dev/fd -> /proc/self/fd
/var/spool/cron/crontabs -> /etc/crontabs
/var/spool/mail -> /var/mail
```

```bash
# bypass `Read sensitive file trusted after startup`
$ cat /var/spool/cron/crontabs/../shadow
$ cat /proc/self/root/etc/shadow
```

and there would be more symlink files in other famaous base images.

So, adding all symlink files into denylist is difficut, we need fundamental solution of detecting this bypass.

The cause of this bypass is **falco do NOT get realpath of syscall argument**.

For example, looking at the command `cat /var/spool/cron/crontabs/../shadow`, a syscall `openat` is executed with argument `/var/spool/cron/crontabs/../shadow`, but

- kernel get realpath of symlink file path `/var/spool/cron/crontabs -> /etc/crontabs` and get `/etc/shadow`
- [falco-lib normalize path](https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/utils.cpp#L632) converts into `fd.name = /var/spool/cron/shadow` by only path string

Thus, [falco-lib normalize path](https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/utils.cpp#L632) should be include getting realpath of symlink file path, but it doesn't, so the syscall returns the contents of `/etc/shadow`, but falco records `/var/spool/cron/shadow`.

So, I propose to add getting realpath into [falco-lib normalize path](https://github.com/falcosecurity/libs/blob/master/userspace/libsinsp/utils.cpp#L632), but there is one problem.

The problem is **it is difficult to get realpath of symlink file in CONTAINER**.

Falco runs as a host or node process, so it is easy to get realpath of symlink file in host. (use `realpath` function)

But in conteiner, container filesystem is made of overlayfs and virtualized, it is difficult to get realpath of symlink file in container. To get realpath of symlink file in container, we have two options, one is read overlayfs files from host, and the another one is create agent that run in container.

To summarize the point of this issue,

- symlink file based bypass method is critical for falco detection
- to detect bypass completely, we have to get realpath of syscall argument in falco
  - in host, it is easy to get realpath
  - in container, it is difficult to get realpath because of virtualized filesystem

and we have to discuss about two points

- Are there other fundamental solutions of detecting symlink file based bypass method?
- If we try to get realpath of syscall argument in falco, how can I do?

I'm waiting your comments. Thank you.

**Comment by incertum** [comment]
CC @loresuso 

@hi120ki this is great and very important! Let me also think more. I think robustness to such obfuscation methods is much needed and there are still many other cases as well.

**Comment by FedeDP** [comment]
Moving to next milestone. Something is moving: https://github.com/falcosecurity/libs/issues/1111
/milestone 0.36.0

**Comment by Andreagit97** [comment]
This should be solved by https://github.com/falcosecurity/libs/pull/1300

**Comment by incertum** [comment]
> This should be solved by [falcosecurity/libs#1300](https://github.com/falcosecurity/libs/pull/1300)

For symlinked executables, not yet for files 🙃 and I think this issue was primarily around files. Perhaps for files we should go with @loresuso proposal and add the respective LSM hook https://github.com/falcosecurity/libs/issues/1111#issue-1719277315. That way we can also refactor the current TOCTOU solution and the way we go about logging file opens in general.
